cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        if (task.suspend()) {
+          activeTasks.decrementAndGet();
+        }
+      }
+    }
+
+    /**
+     * Resume a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void resumeScheduleTask(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null || !task.resume()) {
+        return;
+      }
+      if (activeTasks.incrementAndGet() == 1) {
+        // There were no active tasks until then, that means polling the stream was disabled.
+        // We need to check if it is necessary to poll the stream at this time, if the last
+        // notification received was too long ago, or if there is no last seen notification
+        synchronized (lastNotificationLock) {
+          if (lastNotification == null ||
+            (lastNotification.getTimestamp()
+              + TimeUnit.SECONDS.toMillis(pollingDelay) <= System.currentTimeMillis())) {
+            long streamSize = pollStream();
+            lastNotification = new StreamSizeNotification(System.currentTimeMillis(), streamSize);
+          }
+        }
+      }
+      task.received(lastNotification);
+    }
+
+    /**
+     * Delete a scheduling task that is based on the data received by the stream referenced by {@code this} object.
+     */
+    public void deleteSchedule(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+      StreamSizeScheduleTask scheduleTask = scheduleTasks.remove(getScheduleId(programId, programType, scheduleName));
+      if (scheduleTask != null && scheduleTask.isRunning()) {
+        activeTasks.decrementAndGet();
+      }
+    }
+
+    /**
+     * Get the status a scheduling task that is based on the data received by the stream referenced by {@code this}
+     * object.
+     */
+    public ScheduleState scheduleTaskState(Id.Program programId, SchedulableProgramType programType,
+                                           String scheduleName) {
+      StreamSizeScheduleTask task = scheduleTasks.get(getScheduleId(programId, programType, scheduleName));
+      if (task == null) {
+        return ScheduleState.NOT_FOUND;
+      }
+      return task.isRunning() ? ScheduleState.SCHEDULED : ScheduleState.SUSPENDED;
+    }
+
+    /**
+     * @return true if this object does not reference any schedule, false otherwise
+     */
+    public boolean isEmpty() {
+      return scheduleTasks.isEmpty();
+    }
+
+    public Id.Stream getStreamId() {
+      return streamId;
+    }
+
+    @Override
+    public Type getNotificationFeedType() {
+      return StreamSizeNotification.class;
+    }
+
+    @Override
+    public void received(StreamSizeNotification notification, NotificationContext notificationContext) {
+      // We only pass the stream size notification to the schedule tasks if the notification
```

Comment:
We can use it to access datasets within a transaction. But we don't need that functionality here.